flink/src/main/java/org/apache/iceberg/flink/FlinkTypeToType.java

```diff
@@ -0,0 +1,142 @@
+  private int nextId = 0;
+
+  FlinkTypeToType(FieldsDataType root) {
+    this.root = root;
+    // the root struct's fields use the first ids
+    this.nextId = root.getFieldDataTypes().size();
+  }
+
+  private int getNextId() {
+    int next = nextId;
+    nextId += 1;
+    return next;
+  }
+
+  @Override
+  public Type fields(FieldsDataType dataType, Map<String, Tuple2<String, Type>> types) {
+    List<Types.NestedField> newFields = Lists.newArrayListWithExpectedSize(types.size());
+    boolean isRoot = root == dataType;
+
+    Map<String, DataType> fieldsMap = dataType.getFieldDataTypes();
+    int index = 0;
+    for (String name : types.keySet()) {
+      assert fieldsMap.containsKey(name);
+      DataType field = fieldsMap.get(name);
+      Tuple2<String, Type> tuple2 = types.get(name);
+
+      int id = isRoot ? index : getNextId();
+      if (field.getLogicalType().isNullable()) {
+        newFields.add(Types.NestedField.optional(id, name, tuple2.f1, tuple2.f0));
+      } else {
+        newFields.add(Types.NestedField.required(id, name, tuple2.f1, tuple2.f0));
+      }
+      index++;
+    }
+    return Types.StructType.of(newFields);
+  }
+
+  @Override
+  public Type collection(CollectionDataType collection, Type elementType) {
+    if (collection.getElementDataType().getLogicalType().isNullable()) {
+      return Types.ListType.ofOptional(getNextId(), elementType);
+    } else {
+      return Types.ListType.ofRequired(getNextId(), elementType);
+    }
+  }
+
+  @Override
+  public Type map(KeyValueDataType map, Type keyType, Type valueType) {
+    if (map.getValueDataType().getLogicalType().isNullable()) {
+      return Types.MapType.ofOptional(getNextId(), getNextId(), keyType, valueType);
+    } else {
+      return Types.MapType.ofRequired(getNextId(), getNextId(), keyType, valueType);
+    }
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  @Override
+  public Type atomic(AtomicDataType type) {
+    LogicalType inner = type.getLogicalType();
+    if (inner instanceof VarCharType ||
+        inner instanceof CharType) {
+      return Types.StringType.get();
+    } else if (inner instanceof BooleanType) {
+      return Types.BooleanType.get();
+    } else if (inner instanceof IntType ||
+        inner instanceof SmallIntType ||
+        inner instanceof TinyIntType) {
+      return Types.IntegerType.get();
+    } else if (inner instanceof BigIntType) {
+      return Types.LongType.get();
+    } else if (inner instanceof VarBinaryType ||
+        inner instanceof BinaryType) {
+      return Types.BinaryType.get();
+    } else if (inner instanceof FloatType) {
+      return Types.FloatType.get();
+    } else if (inner instanceof DoubleType) {
+      return Types.DoubleType.get();
+    } else if (inner instanceof DateType) {
+      return Types.DateType.get();
+    } else if (inner instanceof TimeType) {
+      return Types.TimeType.get();
+    } else if (inner instanceof TimestampType) {
```

Comment:
Flink doesn't support `TIMESTAMP WITHOUT TIME ZONE`?

Reply:
It looks like the Flink type that most closely matches is [LocalZonedTimestampType](https://javadoc.io/doc/org.apache.flink/flink-table-common/latest/org/apache/flink/table/types/logical/LocalZonedTimestampType.html) because the Iceberg type does not store the original time zone.

Reply:
Read the javadoc again,  Yes you're right.